sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -631,18 +629,10 @@ private static Query translateGqlQuery(String gql, Datastore datastore, String n
                 .apply(ParDo.of(new GqlQueryTranslateFn(v1Options)));
       }
 
-      PCollection<KV<Integer, Query>> splitQueries = inputQuery
-          .apply(ParDo.of(new SplitQueryFn(v1Options, getNumQuerySplits())));
-
-      PCollection<Query> shardedQueries = splitQueries
-          .apply(GroupByKey.<Integer, Query>create())
-          .apply(Values.<Iterable<Query>>create())
-          .apply(Flatten.<Query>iterables());
-
-      PCollection<Entity> entities = shardedQueries
-          .apply(ParDo.of(new ReadFn(v1Options)));
-
-      return entities;
+      return inputQuery
+          .apply("Split", ParDo.of(new SplitQueryFn(v1Options, getNumQuerySplits())))
+          .apply("Reshuffle", Reshuffle.<Query>viaRandomKey())
```

Comment:
If this is used in Streaming, I am not convinced it is update compatible. It might be (since the GroupByKey types are the same) but it still seems risky.

Reply:
I don't think it is used in streaming.

Reply:
Bounded sources can be used in Streaming.

Reply:
Yes, however it seems overwhelmingly unlikely that a streaming job will be updated while reading a bounded collection from Datastore - if it was a long-running job, most likely it has already read the data. A vanishingly small number of streaming pipelines read from Datastore; an also quite small number of streaming pipelines get updated. Just to be sure, I ran a query on internal logs and there are zero streaming pipelines that both read from Datastore AND have been updated.

I also think we shouldn't treat update compatibility as a hard requirement in general: we have quite limited understanding of exactly what changes are update-compatible or not, and we have no tools at all for smoothly making an update-incompatible (e.g. evolving a coder).

I think it's reasonable to treat update compatibility as a best-effort thing: avoid potentially incompatible changes when doing so is easy, and when doing otherwise is likely to damage a significant number of pipelines.

Reply:
In Dataflow, I think Random().viaRandomKey() is update compatible. It does not break exactly-once processing semantics. An element is processed in downstream stage only after it (and its randomly assigned key) is persisted at upstream range.